### ISSUE TITLE
Proposal: pinning namespaces aka --net=container:name, --net=ns:name without container running

### ISSUE THREAD

**Issue by runcom** [docker/docker#12035]
/cc @ibukanov @thockin @jfrazelle

Author: @ibukanov (Most text is taken from his first issue on this)
Implementation and further ideas: @ibukanov @thockin @runcom

Original proposal/issue: #10642
Related: #11777, maybe others

Here I'm trying to draw a big picture of what is needed and how can be achieved.
Of course, your feedbacks on this is strongly required/needed
Sorry in advance if my english is not perfect. I'm not native and I'll try to be
as clear as possible.

NOTE: I might just refer to network namespace here but this applies to all
namespaces (even user?)

PROBLEM
---------------
As of today, for instance, linking two docker containers requires the first one
to be up and running before starting another one with flags that share first
container's namespaces (eg. `--net:container:foo`, even `--link foo`).
In this scenario it'is not possible to restart or stop the first container. Also,
order matters and in a chain of linked containers one has to be really careful
with startup order.
The fact that something has to be running to keep it open means that there's a
risk - if that running process dies for any reason, all of the things that joined
the (net) namespace are in jeopardy and have to be restarted.

SO?
------
Wouldn't be great if all of this just acts like Volumes? (They continue to exist
and can be used even if the container that firstly created them is not running
anymore)
Container FOO "can" just create the namespaces needed to be "shared". Container
BAR or any other just use FOO's namespaces as needed. Better, as `--parent-group`
option, a container can join whatever namespace available.
Wouldn't be nice if `--net=container:foo` would not require foo running? Rather,
docker can create foo's network namespace and allow other containers to join it.
When foo later starts, it would just join that network namespace. Also, when foo
exits and starts again, it should join the original namespace if other container
still use it rather than creating a new one again.

We just got `--parent-cgroup`, which side-steps the "model everything as a
container". If we had `--pid=ns:name`, `--net=ns:name`, `--ipc=ns:name` we could
do "pinned" namespace management outside of docker too.

QUESTIONS AND POSSIBLE SOLUTIONS (? NEED FEEDBACKS)
-----------------
Given agreements on the above:

1) UX
--------

@thockin gave some hints on how namespaces could be pinned from a container with
the help of a new flag called '--pin-namespaces':

This would pin all container's namespaces:

```
$ docker run -d --pin-namespaces dockerfile/noop
123456789abcdef


# this would normally throw an error if the first container exited!!!
$ docker run --net=container:123456789abcdef my/realjob
8675309deadbeef
```

This way, the first command will just pin its namespaces without
running (I'll explain later what I mean by 'without running') and the
second container will join first one's net namespace.

Also, what about specifying which namespace to save:

```
$ docker run -d --pin-namespaces=net dockerfile/noop
123456789abcdef

$ docker run --net=container:123456789abcdef my/realjob
8675309deadbeef
```

Same as above, but only net namespace is pinned for later use.

2) how are namespaces pinned from containers?
-----------------------------------------------------------------

Pinning container's namespaces requires firstly to have a process running
so it can `setns`ing the namespaces it need.
Now `/proc/{pid}/ns` is populated (initialized). How to pin them and let the
container exit without erasing its namespaces?

- file descriptors to its namespaces should be held before exiting. This will
preserve them
- bind mounting its namespaces somewhere

First option is a little tricky to me and require holding a pointer indefinitively
(how long? where? what if daemon is restarted?).

Second one just requires starting a container just to have its `/prod/{pid}/ns`
populated and mount bind each file somewhere like `/var/lib/docker/{container}/ns`.
However I'm not sure if it's hacky to just run a container to have its namespaces.
How then? just running it with something like `# true` or `# sleep 100`?

```
# docker run -d busybox sleep 100
  0c9c7bee7196845688d8de6a81f49201f3185263a3ca54dcea30b33ee844e83c
# docker inspect --format="{{ .State.Pid }}" 0c9c7bee7196
  24024
# mkdir -p /var/lib/docker/ns/0c9c7bee7196
# touch /var/lib/docker/ns/0c9c7bee7196/net
# mount --bind /proc/24283/ns/net /var/lib/docker/ns/0c9c7bee7196/net
# docker stop 0c9c7bee7196845688d8de6a81f49201f3185263a3ca54dcea30b33ee844e83c
# cd /var/lib/docker/ns/0c9c7bee7196 && nsenter --net=net
# echo $$
  24587
# #inspecting stat of /proc/24587/ns/net show it's using ns from the first stopped container
```

Either case require working on libcontainer too (to my knowledge)

3) how are pinned namespaces garbage collected? Do they?
---
What if the first container is started?
I think it will just look if he has his namespaces pinned somewhere and STILL used,
otherwise it will just create new ones and join them. (This may conflict)

What if the first container is never started and the second one exits?
Pinned namespaces from the first container should be garbage collected (? how, do they?)

Final thoughts
---------------------
This, as said before, would open the possibility to even go further and let
containers use whatever namespace they want (`--net:ns:somensfromnotdockertoo`)
as --parent-cgroup is doing now.

PLEASE COMMENT AND WRITE DOWN YOUR FEEDBACK!
-----------------------------------------------------------------------------------
This is far from a perfect proposal but I'll try to make it better and keep working on it

**Comment by phemmer** [comment]
I've run into a few scenarios where this would be useful. I know kubernetes would find this useful as they start network namespace containers, leave a process in them to keep them alive, and then attach other containers to that one's namespace.

I personally don't like the whole "pin" thing. I don't know how the docker code looks like in this area, but Ideally I would think the following to be the best solution:

* Create the parent namespace container (normal container creation, nothing special).
* Create the child container and link to the first container's namespace (choose whatever namespaces you want).
* Start the child container.
 * If the parent container's namespaces are not active, activate them. (not necessarily launch a process, just initialize the namespaces).
 * Increment a reference counter on the parent's namespace.
* Stop the child container.
 * Decrement the reference counter on all linked namespaces.
   * If any counter is 0, shut down the namespace.

This method would require no changes to the API. From the docker client's perspective it would "just work".

**Comment by LK4D4** [comment]
I like ns:path, but not a fan of pin-namespace. We had idea for libcontainer to have all namespaces pinned by default(apart from PID), but in docker and optional it will be a mess.
Ref counter okay idea as for me. But still in libcontainer, not in docker.

**Comment by thockin** [comment]
@phemmer I think the thing you're glazing over is "not necessarily launch a process, just initialize the namespaces" and "Increment a reference counter on the parent's namespace".  AFAIK there is no way to create a namespace without putting a process in it and no way to keep a namespace open without having a process in it OR bind mounting the FD.  The former is the problem this proposal seeks to avoid, the latter is the thing that you don't like.

Let's think for a minute about the behavior we want, and see if we can build a model of what we NEED vs WANT.  Focusing just on netns, but the same behaviors roughly map to all namepsace types.

Behavior 1.  No two running child containers (defined as using `--net-container:<parent-id>`) of the same parent can ever find themselves in different namespaces from each other.  This does not imply that the namespace exists if there are 0 children running.

Behavior 2.  Consider an only-child container of a parent.  When the child exits and is restarted under the same parent it finds itself with the same IP address.  This does not imply that the namespace exists if there are 0 children running.

Behavior 3.  As in behavior 2, but the restarted child finds all network tunables (iptables rules, sysctls, etc) as they were before the crash.

Behavior 4.  As in behavior 2, but the restarted child finds all network stats as they were before the crash.


Now, I will argue that behavior 1 is a clear NEED.  I'll also posit that behavior 3 is a NEED.  Behavior 2 is WANT but not need, though it makes any implementation of behavior 3 easier.  Behavior 4 is a WANT.


Assuming nobody disagrees with those, let's talk about ways to achieve them.

Pinning namespaces (whether you pin all of them or just ones that have been requested) is the easiest way to implement all of these behaviors.  It has very obvious lifecycle semantics  -- namespaces go away when a container is `docker rm`ed.  Docker itself can observe the lifecycle of the parent.

You could, as @phemmer describes, let the natural process refcount control lifecycle.  That can deliver behavior 1 (a NEED) pretty easily.  Getting behavior 3 (the other NEED) is going to be tricky and error prone (I'd call it a fool's errand - there are an infinite number of things to checkpoint and restore).

You could punt the whole thing to the orchestration layer and say "here's your escape valve: `--net=ns:path-to-something`".  This provides all of the behaviors.  It has pretty clear lifecycle semantics  -- the orchestrator controls lifecycle.  Docker itself can not observe the lifecycle.


I think pinning the namespaces is the only sane way to implement this, frankly.  The only question in my mind is whether the lowest container management layer (Docker) wants to be involved in the process or just provide an escape valve.  I don't frankly care which, though there is clear precedent for running a noop container just to manifest some shared thing -- volumes.  It's a bit clunky as a model (IMO, sorry) but it works and it exists.


to @LK4D4 what mess?  I don't think pinning all namespaces is reasonable - they have low overhead, but it's not zero.  I will boldly assert with no data at all that the vast majority of container runs in the world will NOT want to pin namespaces.  But as an optional feature it is optimal and strictly compatible.

**Comment by thockin** [comment]
As another though experiment, consider behavior 3 with regards to mount namespaces or IPC namespaces.  I should be able to create a parent ns, join it with a single child, have that child mount a tmpfs or bind-mount files around or create a shared memory segment (shared /dev/shm is a concern that Docker seems to not handle right now, #12046), then have that child exit and be restarted and find those things still intact.  That's what behavior #3 means.

Ideally behavior 3 is a WANT and not a NEED, but that presumes all applications become self-healing, and that orchestrators configure everything they want in the form of docker flags.  I think it is unrealistic, so I leave it as a NEED.

**Comment by runcom** [comment]
@LK4D4 I think this can be further splitted in a pull request to support `ns:path` for net, ipc, pid as a first improvement, what do you think?

**Comment by runcom** [comment]
also #12067 seems to point here

**Comment by LK4D4** [comment]
@runcom Do you think it's still needed?

**Comment by win-t** [comment]
Hi, I just run into the same problem, but instead of `--net=container:name` or `--net=ns:name`, I think it's better to create network driver.
so the UX probably like this

1. Create network
```
docker network create -d shared_net -o "connect-to=bridge" mynet
```
or
```
docker network create mybridge
docker network create -d shared_net -o "connect-to=mybridge" mynet
```
2. Run container using this network
```
docker run -d --name mysql --net mynet mysql
docker run -d --name myapp --net mynet local/myaap
```

This also solves the problem with Pod concept in Kubernetes, each container in the same pod share same localhost, AFAIK, Kubernetes using `gcr.io/google-containers/pause` to create dummy container and create other containers on the pod and use the network of the dummy container.

**Comment by win-t** [comment]
I think it's still common practice to use `localhost` as endpoint of another service, 
in some design like `etcd`, `consul`, and `zookeper` the client is expected to connect to `localhost`, but as docker philosophy said, we don't want to add that local proxy service to the image (because that is not part of the app stack)

**Comment by bradrydzewski** [comment]
@win-t also see https://github.com/docker/docker/issues/26024

**Comment by win-t** [comment]
@bradrydzewski, wow thanks
I didn't know if this approach is already considered before

**Comment by nunofernandes** [comment]
I'm in favour of having `--net=ns:name` option with the following use-case:

1 - I create a vpn using standard openvpn. Due to the fact that I don't want to leak routes to the main network namespace, I do it in a new network namespace. The openvpn's "up" scripts are in charge of creating the namespace. 
2 - I would like to be able to run containers on that namespace so that they can access the resources using the vpn connection, so for me it makes sense to have something like:
`docker run -ti --rm --net=ns:openvpn <image>`
